Full Store Editing: Technical documentation (#5939)
woocommerce/woocommerce-gutenberg-products-block: docs/templates/README.md

# WooCommerce Block Templates

This page includes documentation related to WooCommerce Block Templates.

## Table of Contents

* [Overview](#overview)
  * [Requirements](#requirements) 
* [Technical Overview](#technical-overview)
  * [The Problem](#the-problem)
  * [The Solution](#the-solution)
* [Some things to be aware of](#some-things-to-be-aware-of)
* [Related Files](#related-files)

## Overview

> At the highest level, WooCommerce Block Templates are a collection of WooCommere core templates that are available in the WordPress Full Site Editing feature introduced as part of WordPress 5.9. Enabling you to customize these templates with blocks via the Site Editor. 

Reviewer: ```suggestion
At the highest level, WooCommerce Block Templates are a collection of WooCommerce core templates that are made available in the WordPress Full Site Editing experience introduced as part of WordPress 5.9, enabling you to customize these templates with blocks via the Site Editor. 
```

> At the highest level, WooCommerce Block Templates are a collection of WooCommerce core templates that are made available in the WordPress Full Site Editing experience introduced as part of WordPress 5.9, enabling you to customize these templates with blocks via the Site Editor. 

Reviewer 2: ```suggestion
WooCommerce Block Templates are a collection of WooCommerce Core templates for the WordPress Full Site Editing experience introduced in WordPress 5.9. You can customize these templates in the Site Editor. 
```

You can read more about the Full Site Editing (FSE) feature [here](https://developer.wordpress.org/block-editor/getting-started/full-site-editing/).

### Requirements

| Software        | Minimum Version  |
|-----------------|------------------|
| WordPress       | 5.9              |
| WooCommerce     | 6.0              |

> | Block theme     | -                |

Reviewer: Perhaps _“Any block theme”_? Is that the official name, by the way? Or perhaps, “block-enabled theme” or “FSE-enabled theme”?

Am not sure what the official name is. But adding “any” would help by not confusing people that “Block theme” is a specific thing.

Author: >But adding “any” would help by not confusing people 

Yep, I agree

>Am not sure what the official name is.

More info [here](https://developer.wordpress.org/block-editor/getting-started/full-site-editing/#block-themes)

> | Any block theme | -                |

Reviewer 2: ```suggestion
| Theme | Any [block theme](https://developer.wordpress.org/block-editor/how-to-guides/themes/block-theme-overview/#what-is-a-block-theme)               |
```

## Technical Overview

### The Problem

> The FSE feature does not accommodate for loading templates from plugins such as WooCommerce (or WooCommerce Blocks). Instead all template files loaded natively with this feature are to be present within the active themes files. This is a problem because we would like WooCommerce specific templates to be compatible with every theme, and to do this we have to provide themes that don't have WooCommerce specific templates with them from our plugin.

Reviewer 2: ```suggestion
Currently, the FSE feature does not accommodate loading block templates from plugins such as WooCommerce (or WooCommerce Blocks). Instead, all template files loaded natively must be present within the active theme's templates folder. When the active block theme doesn't come with WooCommerce specific templates, we want to be able to provide sensible defaults, though.```

### The Solution

> We created a custom solution which involves a handful of files (listed in the below table) that are responsible for both making the templates available as a block template, and rendering it on the front-end.

Reviewer: ```suggestion
We created a custom solution which involves a handful of files (listed in the table below) that are responsible for both making the templates available as block templates, and rendering them on the front-end.
```

Reviewer 2: ```suggestion
To solve the aforementioned problem, we created a custom solution that involves the files listed below. They are responsible for making the templates available as block templates and rendering them on the front-end.
```

> The BlockTemplateController class is primarily responsible for hooking into both WordPress core, and WooCommerce core filters to load WooCommerce [templates](https://github.com/woocommerce/woocommerce-gutenberg-products-block/tree/trunk/templates/templates) in the Site Editor and on the front-end. It is in this class where the majority of the logic is handled.

Reviewer: ```suggestion
The `BlockTemplateController` class is primarily responsible for hooking into both WordPress core, and WooCommerce core filters to load [WooCommerce templates](https://github.com/woocommerce/woocommerce-gutenberg-products-block/tree/trunk/templates/templates) in the Site Editor and on the front-end. It is in this class where the majority of the business logic is handled.
```

### Some things to be aware of

> * Individual templates are represented by a placeholder block within the Site Editor until we can 'block-ify' these. This is the long-term goal. Until then users will be able to customize templates by adding blocks above and below the placeholder.

Reviewer: ```suggestion
* Individual templates are represented in their entirety by a placeholder block within the Site Editor until we can cut them up into discrete blocks. This is our long-term goal. However, until then, users will be able to customize templates by adding blocks above, below and around the placeholder.
```

Reviewer 2: Agree that 'block-ify' is unclear! That said, maybe `divide them up` instead of `cut them up` here? Might just be me but `cut them up` does feel like it could have a bit of a negative connotation. 😅

Also, I'd leave it at above and below, as around isn't really that much of an option in the current state.

> * At the beginning of the project, we unintentionally used the incorrect WooCommerce plugin slug. This has resulted in us maintaining both the incorrect and correct slugs. We reference these via `BlockTemplateUtils::DEPRECATED_PLUGIN_SLUG` and `BlockTemplateUtils::PLUGIN_SLUG`. More information on that [here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5423).

Reviewer 2: ```suggestion
* At the beginning of the project, we unintentionally used an incorrect WooCommerce plugin slug. We thus maintain both the incorrect and correct slugs (referenced as `BlockTemplateUtils::DEPRECATED_PLUGIN_SLUG` and `BlockTemplateUtils::PLUGIN_SLUG`). More information on that [here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/5423).
```

> * If a theme has a `archive-product.html` template file, but does not have any taxonomy related files. We will automatically duplicate their themes `archive-product.html` file in place of these taxonomy files within `BlockTemplateController->get_block_file_template()`.

Reviewer 2: ```suggestion
* If a theme has an `archive-product.html` template file, but does not have any other taxonomy-related files, we will automatically duplicate their theme's `archive-product.html` file in place of these taxonomy files within `BlockTemplateController->get_block_file_template()`.
```

| File                        | Description                                                                                                                                                                     | Source                                                                                                                          | Docs                                                      |
| --------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------- |
| templates/templates/\*      | Location in the filesystem where WooCommerce block template HTML files are stored.                                                                                              | [Source files](https://github.com/woocommerce/woocommerce-gutenberg-products-block/tree/trunk/templates/templates)              |                                                           |

> | legacy-template/\*          | The JavaScript block rendered in the Site Editor. This is a server-side rendered component which is handled by LegacyTemplate.php                                               | [Source file](https://github.com/woocommerce/woocommerce-gutenberg-products-block/tree/trunk/assets/js/blocks/legacy-template)  | [README](../../assets/js/blocks/legacy-template/README.md)|

Reviewer: ```suggestion
| legacy-template/\*          | The JavaScript block rendered in the Site Editor. This is a server-side rendered component which is handled by `LegacyTemplate.php`                                             | [Source file](https://github.com/woocommerce/woocommerce-gutenberg-products-block/tree/trunk/assets/js/blocks/legacy-template)  | [README](../../assets/js/blocks/legacy-template/README.md)|
```

> | classic-template/\*         | The JavaScript block rendered in the Site Editor. This is a server-side rendered component which is handled by ClassicTemplate.php                                              | [Source file](https://github.com/woocommerce/woocommerce-gutenberg-products-block/tree/trunk/assets/js/blocks/classic-template)  | [README](../../assets/js/blocks/classic-template/README.md)|

Reviewer 2: ```suggestion
| classic-template/\*         | The JavaScript block rendered in the Site Editor.                                              | [Source file](https://github.com/woocommerce/woocommerce-gutenberg-products-block/tree/trunk/assets/js/blocks/classic-template)  | [README](../../assets/js/blocks/classic-template/README.md)|
```

Reviewer 2: This feels redundant as we're addressing ClassicTemplate.php below.

> | ClassicTemplate.php         | Class used to setup the block on the server-side and render the correct template                                                                                                | [Source file](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/src/BlockTypes/ClassicTemplate.php) | [README](./classic-template.md)                            |

Reviewer 2: ```suggestion
| ClassicTemplate.php         | Class used to setup the block on the server-side and render the correct template.                                                                                                | [Source file](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/src/BlockTypes/ClassicTemplate.php) | [README](./classic-template.md)                            |
```

> | BlockTemplateController.php | Class which contains all the business logic which loads the templates into the Site Editor or on the front-end through various hooks available in WordPress & WooCommerce core. | [Source file](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/src/BlockTemplatesController.php)  | [README](./block-template-controller.md)                  |

Reviewer 2: ```suggestion
| BlockTemplateController.php | Class containing all the business logic loading the templates into the Site Editor or on the front-end through various WordPress & WooCommerce core hooks. | [Source file](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/src/BlockTemplatesController.php)  | [README](./block-template-controller.md)                  |
```

| BlockTemplateUtils.php      | Class containing a collection of useful utility methods.                                                                                                                        | [Source file](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/src/Utils/BlockTemplateUtils.php)  |                                                           |

> 

Reviewer: Should all the elements within the first column be surrounded by backticks perhaps?
